### PR TITLE
Remove dependency on apr by using c++11 / c++14 features

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We had some limited use of apr left in our tcnative fork which complicated the deployment. This apr usages can be replaced by using c++.

Modification:

Remove the dependency on apr by using c++11 / c++14 features

Result:

No apr dependeny

### CODE REVIEW

**Comment by normanmaurer** [comment]
Still some cleanup missing

**Comment by normanmaurer** [comment]
This is ready to review now! :)

File: openssl-dynamic/src/main/c/tcn_lock.h
Comment (Scottmitch) [inline openssl-dynamic/src/main/c/tcn_lock.h:25]: nit: use `malloc`/`free` terminology or `new`/`delete`, since this is a `c` API maybe the former makes sense?

**Review by Scottmitch — COMMENTED** [review]
good work! took a quick peak at c API and c++ bridge code and lgtm ... will defer to someone else to take a deeper look.

File: openssl-dynamic/src/main/c/tcn_lock_rw.cpp
Comment (normanmaurer) [reply openssl-dynamic/src/main/c/tcn_lock_rw.cpp:21]: maybe we can even use c++17... let me double check

**Comment by chrisvest** [comment]
I don't think we need to add C++ to the mix. C11 has atomics and mutexes as well: https://en.cppreference.com/w/c/thread

**Review by chrisvest — APPROVED** [review]

**Comment by normanmaurer** [comment]
@chrisvest I tried this after reading your comment but it turns out not a lot of glibc / compilers implement it (yet). See also https://www.reddit.com/r/C_Programming/comments/v0xxnj/is_c11_threadsh_worth_using/. So maybe lets just stick to c++11 etc

File: openssl-dynamic/src/main/c/tcn_lock_rw.cpp
Comment (normanmaurer) [reply openssl-dynamic/src/main/c/tcn_lock_rw.cpp:21]: yeah seems like we not always have it... so lets just leave it as it is

**Comment by normanmaurer** [comment]
I am still investigating some linking stuff. Dont merge yet

**Comment by chrisvest** [comment]
Alright. Ping me when it's ready to be looked at again.

**Review by chrisvest — APPROVED** [review]

**Pull request merged**
